### ISSUE TITLE
SSR: class order is not persisted

### ISSUE THREAD

**Issue by Filipeue** [styled-components/styled-components#3321]
SSR + styled-components:
```
const A = styled.div`
  position: relative;
`
const B = styled(A)`
  position: absolute;
`
```
It happens that element B has `position: relative` (the order of generated classes is wrong) from time to time. It only happens when using SSR and not on all elements... pretty weird.... Has anyone exprerienced the same isssue?
I can create proper issue using the bug template if necessary.

**Comment by marhalpert** [comment]
I have the same problem when migrating from 5.1.1 to 5.2.0.

SRR looks good, then the app re-rendered and wrong styles get applied.

```jsx
${({ $block }) => $block && css`
  display: block;
  width: 100%;

  & + & {
    margin-top: ${rem(10)};
  }
`}
```

Even with the second button is getting the margin-top

```jsx
   <Button block={false} /> 
   <Button block={false} /> 
```

**Comment by nesbtesh** [comment]
Samething is happening to me, its becoming a big issue. On my app I thaught it was loadable components. See issue: https://github.com/gregberge/loadable-components/issues/656

But after tracing it, I think its an issue with styled-component

**Comment by nesbtesh** [comment]
Just for debugging purposes I removed the ServerStyleSheet generator from the server and it works fine once everything renders in the front end. So the bug must be in the server style sheet manager

**Comment by Filipeue** [comment]
it is issue even in v 4.4.1, and it is becoming serious issue

**Comment by pajasevi** [comment]
I am encountering this issue in v5.2.1 but not in v5.2.0.

**Comment by rmoedt** [comment]
Same for me, 5.2.1 has the issues, 5.2.0 is fine.

**Comment by antoinerousseau** [comment]
Same here, issue happens with 5.2.1 but not 5.2.0.

https://github.com/styled-components/styled-components/compare/v5.2.0...v5.2.1#diff-a92a3ff385b13e270902408a5eca54cbd783bba20b200e5bbc62e957dc127faf

**Comment by antoinerousseau** [comment]
I am able to reproduce it in a non-random way:
- Go to https://5fad74a59bb5d60008255c2d--agencephare.netlify.app/agenda/
- Open the "agence" menu and click the "agence" entry
On that "agence" page, the big blue block (starting with "Depuis 2014...") has a `max-width: 900px` instead of `1200px`

Now reload this "agence" page (Cmd-R): the block is now 1200px wide as expected.

The block is a `Section` and here is the related code:

```js
export const LAYOUT = {
  maxWidth: 1200,
  column: 900,
  //...
}

export const Column = styled.section`
  max-width: ${LAYOUT.column}px;
  margin: 0 auto;
  padding: ${LAYOUT.margin}px;
`
export const Section = styled(Column)`
  max-width: ${LAYOUT.maxWidth}px;
`
```

**Comment by tbekaert** [comment]
I have the same issue and I think it is caused by this PR https://github.com/styled-components/styled-components/pull/3233

@kitten @probablyup @mu29 Hi guys, as you worked on that PR, do you think it could comes from it?

**Comment by Filipeue** [comment]
Even downgrading to 5.2.0 did NOT help... I m becomming desperate :( 
Anyone has minimal codepen setup for styled-components SSR?

**Comment by Filipeue** [comment]
I have located the issue to the babel plugin `babel-plugin-styled-components`.
Downgrading the plugin to `1.10.7` solved the issue!

**Comment by antoinerousseau** [comment]
Did you open an issue at https://github.com/styled-components/babel-plugin-styled-components/issues ?

**Comment by kitten** [comment]
@tbekaert That's a very astute observation. I made this comment back then and I believe that could be what's happening here. Since the PR in question doesn't actually fix anything we can probably just revert it. Have you by any chance tried that and built a preview version?

>  I still think we have to further test and think whether it holds true though, since SSR results may be very much out of order, if preregistration fails.

I believe I simply never quite got to testing this before this was already pushed out.

Since no one here seems to have a reproduction (@antoinerousseau A production site with just a link is not a reproduction, it's just showing me that either you or we made a mistake) that'd probably be able to confirm that this is the issue.

@Filipeue So this is where it gets interesting, I _do_ believe that #3233 could be causing some issues. But I'm _assuming_ that people here have correct setups. If your setup is broken then that'd obviously entirely unrelated.

Since this happens in a lot of our issues it's almost impossible to see which issues are relevant or not. So please refrain from comments just because you're desperate for a fix. They don't move this along 😅 

So tl;dr: @tbekaert @antoinerousseau You seem to have solid cases that can reproduce the bug. Would it help if I provided a pre-release with this PR reverted? I'd also love to know more about your SSR setup to make sure that it's not causing the issue through other means 🙌

**Comment by antoinerousseau** [comment]
@kitten the website mentioned at https://github.com/styled-components/styled-components/issues/3321#issuecomment-726267797 uses [this Gatsby TS template](https://github.com/antoinerousseau/gatsby-starter-antoine/tree/netlify) and is built and served by Netlify.

Notable packages used:

- TypeScript
- styled-components
- [babel-plugin-styled-components](https://github.com/styled-components/babel-plugin-styled-components)
- [gatsby-plugin-styled-components](https://www.gatsbyjs.com/plugins/gatsby-plugin-styled-components/)

**Comment by Seresigo** [comment]
Same problem https://github.com/styled-components/styled-components/issues/3207

**Comment by iDVB** [comment]
We have been wrestling with what sounds like this exact same issue. 
Downgrading to SC v5.2.0 fixes it for us. 
This makes little sense, if your change logs are accurate though, since 5.2.1 looks like only [3 small unrelated commits](https://github.com/styled-components/styled-components/releases/tag/v5.2.1) if we are reading this right?   

We have [reproduced the issue in a demo](https://github.com/iDVB/sc-test), BUT its super odd and requires the use of an external lib that has a peer dependency on SC. 
Since our lib is private, I can't really provide that, but happy to talk thru it if this example is at all helpful.

**Repo Steps:**
1. `yarn build && serve`
2. Browse to Home page (index.js)
3. Click the link named "link".  This should link you to Page 2.
4. When you get here... notice the background of the text "Page 2" will be RED
5. Hard refresh the page and watch the text background change to Green.

**EXPECTED:**
The background of the text should be GREEN regardless if navigated to OR hard refreshing.

**ACTUAL:**
The background of the text is RED. until hard refeshing


The only code of note in this base gatsby install repo is...

**/gatsby-ssr.js**
Notice here we only simply only importing from our external lib (with peer dep on SC) and not even using it.
Also, using `setPostBodyComponents` to inject meaningless div. 
Removing EITHER of these will fix the issue.
```javascript
import { consoleKlickSignature } from '@klickmarketing/react-components'

const React = require('react')

export const onRenderBody = ({
  setBodyAttributes,
  setPostBodyComponents,
  setHeadComponents,
}) => {
  setPostBodyComponents([<div>hello</div>])
}
```

**/P.js**  (Custom Paragraph Component)
```javascript
import styled from 'styled-components'

const P = styled.p`
  background: red;
`

export { P }
```

**/index.js**  (HOME PAGE)
```javascript
import React from "react"
import styled from 'styled-components'
import { P } from '../components/P'
import { Link } from "gatsby"

const IndexPage = () => (
  <div>
  <Link to="/page-2">link</Link>
  <OtherP>Page 1</OtherP>
</div>
)

const OtherP = styled(P)`
  background: green;
`

export default IndexPage
```

**/page-2.js**   (PAGE 2)
```javascript
import React from "react"
import styled from 'styled-components'
import { P } from '../components/P'

const SecondPage = () => (
  <div>
  <StyledP>Page 2</StyledP>
</div>
)

const StyledP = styled(P)`
  background: green;
`

export default SecondPage
```

**Comment by gavinsharp** [comment]
> Would it help if I provided a pre-release with this PR reverted?

@kitten I think I'm seeing this bug (class ordering not consistent; working in 5.2.0, broken in 5.2.1), though it's happening on the client-side, not SSR. I am having a lot of trouble creating a minimal test case (our app is complicated), so I tried building a custom package with only #3233 reverted, but ran into some issues getting that version to work. It would be awesome if you could provide a pre-release with that revert, and I could test it quickly to confirm.

**Comment by Filipeue** [comment]
for anyone still dealing with the issue, my findings are that class order is not always persisted and you should not rely on it being persisted
instead you can safely override style like this:
```typescript
const A = styled.div`
  color: black;
`
const B = styled(A)`
  && {
    color: red;
  }
`
const C = styled(B)`
  &&& {
    color: green;
  }
`
```
this way A will be black, B red and C green

**Comment by pajasevi** [comment]
@Filipeue well that's good but it's still an ugly hack for an issue that should not exist. 🤷

**Comment by Filipeue** [comment]
> @Filipeue well that's good but it's still an ugly hack for an issue that should not exist. 🤷

I couldn't agree more, that's why I didn't close the issue

**Comment by migueloller** [comment]
Just wanted to add to the conversation that we were running into this issue as well on `5.2.1` and were able to fix it by downgrading to `5.2.0`. Here's a screenshot of the SSR result:

![image](https://user-images.githubusercontent.com/5677929/113778138-0c764500-96fa-11eb-94c5-a95fb5290e29.png)

And then, when a prop changed in the client, it would result in the following:

![image](https://user-images.githubusercontent.com/5677929/113778180-18620700-96fa-11eb-9ae6-035ad6168730.png)


As can be seen, `ckDfJz` jumps to the top and replaces all the values within the media queries. The styles in `ckDfJz` are in a component that is composed. i.e.,

```typescript
const BackgroundsContainer = styled.div`
  /* uses class ckDfJz, which is stable because the rules don't change */
  position: relative;
  width: 100%;
  margin: 0px auto;
`

const Box = styled(BackgroundsContainer)`
  /* code to compute width and margin, class changes whenever there are different values */
`
```

**Comment by iDVB** [comment]
We just made the switch to Gatsby 3 and we're even getting this now on `styled-components v5.2.0`
It seems to only happen in `yarn build` and not `yarn start`. 

We're also developing with our down component lib and when we `yarn link` that component (which uses styled-components as a peer dep) it also **works**.

I'm going to continue to downgrade various components in hopes that we find another one that is also leading to the isssue.  However, we're pretty in the dark here as to what is causing it.

**Comment by iDVB** [comment]
**UPDATE**
More testing turned up something VERY odd. 

1. I've `ybs` (eg. `yarn build && yarn serve`) and opened in Chrome Incognito
2. I've then clicked around the side and everything looks normal.
3. Then I click on a link to a page I know to be 404. I intensionally, for this test, do NOT have a custom 404 page. so I get an error like this...
![image](https://user-images.githubusercontent.com/189506/114728298-fa556180-9d0c-11eb-8dca-4ccdeceaf80a.png)

4. I then hit the back button and it goes back to the site looking **WORKS**.
5. I then click on ANY working link and the page **WORKS**.
6. I then click on ANY working link and the whole page is **NOT WORKING**.
7. What I'm comparing in the following is the same page when it **WORKS** and **NOT WORKING**.

![image](https://user-images.githubusercontent.com/189506/114729193-c3338000-9d0d-11eb-90fa-635a7380da11.png)

WEIRD PART: Even though this is a completely different module and between this case and the one I [last posted about](https://github.com/styled-components/styled-components/issues/3321#issuecomment-817431012) they **both** use the same styled hash `kiRdWm`. In the first example above it was IN the SSR. In this example here, the code is NOT in SSR, as the component is only conditionally injected. 

We're having a really tough time reproducing this in a small example that we can share since this uses our internal styled-components lib that we can't really make public.

**Comment by iDVB** [comment]
**- UPDATE -**
We now have a reproduction repo(s) for someone to look at should you wish.

Repo Steps:
1. `git clone git@github.com:iDVB/gatsby3_site_issue.git`
2. `yarn install`
3. `yarn build && yarn serve`
4. Browse to `http://localhost/` (Notice cookie banner at bottom looks like this, as expected)
![image](https://user-images.githubusercontent.com/189506/114792961-593fc880-9d57-11eb-8052-cd54f717161d.png)

5. Click `Go to Page 2` link. (Notice the cookie banner looks fine again)
6. Now on this page, hard refresh. 
7. This should now BREAK the cookie banner and it should look like this...
![image](https://user-images.githubusercontent.com/189506/114792994-6d83c580-9d57-11eb-8cb2-f6ec25e0934c.png)


* - Note - Here is the sample styled-components lib repo ([react-components](https://github.com/iDVB/react-components)) if you want to play in that as well.

**Comment by iDVB** [comment]
@kitten @probablyup @mu29 Has anyone had the time to review the above [reproduction repo](https://github.com/styled-components/styled-components/issues/3321#issuecomment-819915948)? ⬆️ 
Seems pretty clear now that this is an issue in both `v5.2.0` or `v5.2.3` and Gatsby3.

We're dead in the water with our projects until we have a work around or fix. Can We help troubleshoot or work on this?

**Comment by iDVB** [comment]
**FIXED**
We managed to find a fix to [our issue](https://github.com/styled-components/styled-components/issues/3321#issuecomment-819915948). It appears this issue has already been identified in [another issue](https://github.com/styled-components/babel-plugin-styled-components/issues/280) on `babel-plugin-styled-components`.

**Comment by iDVB** [comment]
Sigh.... back to this same issue again. This time its even WITH 
`styled-components v5.2.3`
`babel-plugin-styled-components@1.10.5`

However, `styled-components@5.2.3` uses `babel-plugin-styled-components@1.12.0` internally. We tried using a resolution to drop it to match `1.10.5` but then we would get errors.

![image](https://user-images.githubusercontent.com/189506/119545258-9fc61100-bd60-11eb-8924-929cc86baa63.png)

**Comment by mnajdova** [comment]
This surfaced as an issue as well when using in integration with `nextjs` and @material-ui/core which uses `styled-components` for the core components. 

Here is the issue: https://github.com/mui-org/material-ui/issues/27412 and this is the repository that showcase the problem - https://github.com/mnajdova/27412-nextjs-with-styled-components. 

When using `styled` from `styled-components` for overriding the styles of some Material-UI component, it works on the first load and then breaks consistently when refreshing the page. I am using `@latest` version of everything in the example repository, but I can lock the versions if necessary.

**Comment by kitten** [comment]
Duplicate of #3482

We had some other reports of this, I believe, but between the noise it's getting very hard to enter discussions like these, so hence I'm taking a _newer_ issue to address this. Sorry!
